### PR TITLE
Requestless beacon update

### PR DESCRIPTION
**Pull request by bbenligiray**
It feels like we really don't want to wait for PSP to be able to update beacons with signed data so I added an additional function for this. You can also think of this as a new and primitive protocol.

So the Airnode obviously signs `templateId` and `parameters` to signify that this specific request is what it's responding and not something else. We could have left it at that, but it's possible that the Airnode operator may not want to sign such requests because they would have no control over how the response is monetized. Since we ended up with authorizing requester addresses (and built the entire monetization scheme around that), it felt like a good idea to also add the address of the contract that will be ingesting the response (and chain ID is added for good measure). Then, the Airnode may choose to only sign responses that can be used by a specific RrpBeaconServer for example.

It feels like we may want to reuse this pattern in future contracts too, so instead of embedding the signature verification bit in RrpBeaconServer, I went with exposing it at AirnodeRrp.

### CODE REVIEW

**Comment by bbenligiray** [comment]
There is a good argument for not including `block.chainid` and `address(this)` to the request ID. Say I own a contract that the respective Airnode operator does not want to serve. This Airnode operator does want to serve an RrpBeaconServer out there though. I can simply hardcode the chain ID and address of the trusted RrpBeaconServer in my contract code (instead of referring to `block.chainid` and `address(this)`) and fool the Airnode operator into signing a response for me. So since adding these to the request ID hash doesn't necessarily provide control over who can ingest the data, it may be a good idea to omit this.

**Comment by acenolaza** [comment]
> There is a good argument for not including `block.chainid` and `address(this)` to the request ID. Say I own a contract that the respective Airnode operator does not want to serve. This Airnode operator does want to serve an RrpBeaconServer out there though. I can simply hardcode the chain ID and address of the trusted RrpBeaconServer in my contract code (instead of referring to `block.chainid` and `address(this)`) and fool the Airnode operator into signing a response for me. So since adding these to the request ID hash doesn't necessarily provide control over who can ingest the data, it may be a good idea to omit this (so `requestId` in `updateBeaconWithoutRequest()` would just be `beaconId`).

So I guess the question is: do we need Airnode to have control over who consumes the signed data? If yes, then the only viable solution seems to be to have some access control on the Airnode endpoint that will be providing the signed data.

**Comment by bbenligiray** [comment]
> If yes, then the only viable solution seems to be to have some access control on the Airnode endpoint that will be providing the signed data.
We already have that in the form of https://docs.api3.org/airnode/v0.3/reference/deployment-files/config-json.html#httpgateway-apikey so I guess I was trying to provide more granular control, which now doesn't look possible. You just have to make sure that whoever has that API key is not doing "unauthorized" things (whatever that is). Actually this is how all web2 APIs work anyway (you give your user an API key and tell them to abide by the T&C or you will sue them or something) so it feels acceptable.

File: packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol [hunk 1]
```diff
@@ -253,7 +253,13 @@ contract AirnodeRrp is
...
             "Invalid request fulfillment"
         );
-        verifySignature(requestId, airnode, data, signature);
+        require(
+            (
+                keccak256(abi.encodePacked(requestId, data))
+                    .toEthSignedMessageHash()
+            ).recover(signature) == airnode,
+            "Signature mismatch"
+        );
```
Comment (acenolaza) [inline packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol:262]: Shouldn't we have an internal pure `verifyHash(...)` function that gets called here and in `verifyData()` (329-335)

**Review by acenolaza — APPROVED** [review]
LGTM

**Comment by aquarat** [comment]
I was a bit surprised at the hard-coded 1 hour limits; I would have thought this would be up to consuming contracts to decide or perhaps a configurable value on the contract.

**Comment by acenolaza** [comment]
> I was a bit surprised at the hard-coded 1 hour limits; I would have thought this would be up to consuming contracts to decide or perhaps a configurable value on the contract.
> 
> I suppose the future-dated check is to prevent someone locking the contract's update-ability in future.

https://github.com/api3dao/airnode/pull/695#discussion_r757507926

**Comment by bbenligiray** [comment]
> I was a bit surprised at the hard-coded 1 hour limits; I would have thought this would be up to consuming contracts to decide or perhaps a configurable value on the contract.

I avoided this because as soon as you make it configurable, you will need to read it from the storage whenever you need it (i.e., every fulfillment transactions), so it's a good practice to hardcode anything that you can. I couldn't think of a legitimate reason to change this value so I thought this was an acceptable optimization.

> I suppose the future-dated check is to prevent someone locking the contract's update-ability in future.

Right, either accidentally (through a bug) or on purpose

**Review by Ashar2shahid — APPROVED** [review]
Contract code LGTM, reviewing tests but giving the approval if we want to get this merged quickly

File: packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol (hunk 1 above)
Comment (bbenligiray) [reply packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol]: Done here 27ddc8907cdde73fab51201727c25afb5abc48b1

File: packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol [hunk 2]
```diff
@@ -256,7 +258,7 @@ contract AirnodeRrp is
                 keccak256(abi.encodePacked(requestId, data))
                     .toEthSignedMessageHash()
             ).recover(signature) == airnode,
-            "Invalid signature"
+            "Signature mismatch"
```
[the remark this answers is not in the document]
Comment (bbenligiray) [reply packages/airnode-protocol/contracts/rrp/AirnodeRrp.sol]: The point is that it's also possible for the signature to be invalid (by having the wrong length for example) so I wanted to differentiate here

File: packages/airnode-protocol/contracts/rrp/requesters/RrpBeaconServer.sol [hunk 3]
```diff
@@ -246,11 +246,58 @@ contract RrpBeaconServer is
...
             data,
-            (int256, uint256)
+            signature
         );
+        (int256 decodedData, uint256 decodedTimestamp) = ingestData(
+            beaconId,
+            data
+        );
+        emit UpdatedBeaconWithoutRequest(
+            beaconId,
```
[the remark this answers is not in the document]
Comment (bbenligiray) [reply packages/airnode-protocol/contracts/rrp/requesters/RrpBeaconServer.sol:284]: I encounter this predicament a lot while working on protocol-v1. My general approach is that if you're doing off-chain analysis, you can afford to make a few more API calls.

**Pull request merged**